servers/src/main/java/tachyon/worker/block/evictor/EvictorBase.java

```diff
@@ -0,0 +1,200 @@
+    candidateDirView = dirCandidates.candidateDir();
+    List<Long> candidateBlocks = dirCandidates.candidateBlocks();
+    StorageTierView nextTierView = mManagerView.getNextTier(candidateDirView.getParentTierView());
+    if (nextTierView == null) {
+      // This is the last tier, evict all the blocks.
+      plan.toEvict().addAll(candidateBlocks);
+      for (Long blockId : candidateBlocks) {
+        try {
+          BlockMeta block = mManagerView.getBlockMeta(blockId);
+          if (null != block) {
+            candidateDirView.markBlockMoveOut(blockId, block.getBlockSize());
+          }
+        } catch (NotFoundException nfe) {
+          continue;
+        }
+      }
+    } else {
+      for (Long blockId : candidateBlocks) {
+        try {
+          BlockMeta block = mManagerView.getBlockMeta(blockId);
+          if (null == block) {
+            continue;
+          }
+          StorageDirView nextDirView =
+              mAllocator.allocateBlockWithView(Users.MIGRATE_DATA_USER_ID, block.getBlockSize(),
+                  BlockStoreLocation.anyDirInTier(nextTierView.getTierViewAlias()), mManagerView);
+          if (nextDirView == null) {
+            nextDirView =
+                cascadingEvict(block.getBlockSize(),
+                    BlockStoreLocation.anyDirInTier(nextTierView.getTierViewAlias()), plan);
+          }
+          if (nextDirView == null) {
+            // If we failed to find a dir in the next tier to move this block, evict it and
+            // continue. Normally this should not happen.
+            plan.toEvict().add(blockId);
+            candidateDirView.markBlockMoveOut(blockId, block.getBlockSize());
+            continue;
+          }
+          plan.toMove().add(
+              new Pair<Long, BlockStoreLocation>(blockId, nextDirView.toBlockStoreLocation()));
+          candidateDirView.markBlockMoveOut(blockId, block.getBlockSize());
+          nextDirView.markBlockMoveIn(blockId, block.getBlockSize());
+        } catch (NotFoundException nfe) {
+          continue;
+        }
+      }
+    }
+
+    return candidateDirView;
+  }
+
+  @Override
+  public EvictionPlan freeSpaceWithView(long bytesToBeAvailable, BlockStoreLocation location,
+      BlockMetadataManagerView view) {
+    mManagerView = view;
+
+    List<Pair<Long, BlockStoreLocation>> toMove = new ArrayList<Pair<Long, BlockStoreLocation>>();
+    List<Long> toEvict = new ArrayList<Long>();
+    EvictionPlan plan = new EvictionPlan(toMove, toEvict);
+    StorageDirView candidateDir = cascadingEvict(bytesToBeAvailable, location, plan);
+
+    mManagerView.clearBlockMarks();
+    if (candidateDir == null) {
+      return null;
+    }
+
+    return plan;
+  }
+
+  /**
+   * @return an iterator over the blocks in the evictor cache. The key of the map entry is the
+   * block Id.
+   */
+  protected abstract Iterator<Map.Entry<Long, Object>> getIterator();
+
+  /**
+   * Perform additional cleanup when a block is removed from the iterator.
```

Comment:
Could be more specific with `from the iterator returned by {@link getIterator}`.